esp-hal/src/rmt.rs

```diff
@@ -1489,17 +1497,49 @@ pub enum LoopCount {
     /// Repeat until explicitly stopped.
     Infinite,
 
+    // FIXME: Does continuous tx trigger the End interrupt on each repetition such that it could
+    // be used to emulate the LoopCount interrupt on devices that lack it?
     /// Repeat the given number of times.
-    Finite(NonZeroU16),
+    #[cfg(rmt_has_tx_loop_count)]
+    Finite(core::num::NonZeroU16),
+}
+
+impl LoopCount {
+    // Get value to write to hardware (should compile to a no-op)
+    #[allow(dead_code)]
+    fn get(self) -> u16 {
+        match self {
+            #[cfg(rmt_has_tx_loop_count)]
+            LoopCount::Finite(repeats) => repeats.get(),
+            LoopCount::Infinite => 0,
+        }
+    }
 }
 
 // LoopCount::Infinite should be represented as 0u16, which corresponds to the value that needs to
 // be written to the tx_lim register.
+#[cfg(rmt_has_tx_loop_count)]
 const _: () = if core::mem::size_of::<LoopCount>() != 2 {
     // This must not be defmt::panic!, which doesn't work in const context.
     core::panic!("Niche optimization unexpectedly not working!");
 };
 
+/// Stopping behavior for continuous transmission.
+///
+/// Depending on hardware support, automatic stopping of the transmission upon reaching the
+/// specified loop count may not be available.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum LoopStop {
```

Comment:
Don't these options conflict with LoopCount? Manual + Infinite, Auto + Finite seem very confusing combinations to me

Reply:
Hm I'm not sure about NonZeroU16, perhaps we can just return immediately with success when someone wants 0 loops. NonZero just makes things annoying.

Otherwise I'm in favour of the single enum, it's much easier to understand.

Reply:
>  Infinite(Option<NonZeroU16>

that's worse, not better :D

Reply:
:laughing: 

I'll try to go for 

```rust
enum LoopMode {
    /// Infinite, no interrupt enabled
    // Previously LoopCount::Infinite
    Infinite,

    /// Infinite, but sets the interrupt flag after the given number of repetitions
    // Previously LoopCount::Finite + LoopStop::Manual
    #[cfg(rmt_has_tx_loop_count)]
    InfiniteWithInterrupt(u16),

    /// Stops after the given number of iterations; not supported on all devices
    // Previously LoopCount::Finite + LoopStop::Auto
    #[cfg(rmt_has_tx_loop_auto_stop)]
    Finite(u16),
}
```
with loop count 0 returning immediate success and see how that works out. The `NonZero` came about to get the niche optimization in the original `LoopCount` enum, but that's probably not a big deal.

Reply:
you don't really gain much with niche optimization here - if you have 3 options you need 2 bits of niche anyway, but this is a value that will more than likely be const-propagated if the function itself is inlined. It's better to optimize for less boilerplate than one word that may or may not actually be saved.

Reply:
Sorry, I was unclear, I was talking about the previous

```rust
enum LoopCount {
    Infinite,
    Finite(NonZeroU16)
}
```

which had the added advantage that the representation of this enum was exactly the value to write to the hw registers. As you say, with the redesign, niche optimization won't work, but in the end, the advantage of having that optimization in the first place should be minimal.